File: backend/gamification.py
"""Volunteer levels (§28 gamification, extension).

A level is computed on the fly from the same aggregates the stats endpoint
already collects — no extra tables. Points blend deliveries and rescued kg
so both the «many small trips» and the «few heavy hauls» play styles level up.
"""
from typing import Any, Dict

POINTS_PER_DELIVERY = 10.0
POINTS_PER_KG = 1.0

# (code, points threshold) — ordered ascending. i18n labels live on the frontend.
LEVELS = [
    ("novice", 0),        # Новичок
    ("helper", 50),       # Помощник
    ("courier", 200),     # Курьер добра
    ("guardian", 600),    # Хранитель района
    ("city_hero", 1500),  # Герой города
]


def compute_points(deliveries: int, kg: float) -> float:
    return deliveries * POINTS_PER_DELIVERY + kg * POINTS_PER_KG


def compute_level(deliveries: int, kg: float) -> Dict[str, Any]:
    """Current level + progress towards the next one (for the progress bar)."""
    points = compute_points(deliveries or 0, float(kg or 0))
    current = LEVELS[0]
    nxt = None
    for i, (code, threshold) in enumerate(LEVELS):
        if points >= threshold:
            current = (code, threshold)
            nxt = LEVELS[i + 1] if i + 1 < len(LEVELS) else None
    if nxt is None:
        progress = 1.0
        to_next = 0.0
    else:
        span = nxt[1] - current[1]
        progress = (points - current[1]) / span if span > 0 else 1.0
        to_next = max(0.0, nxt[1] - points)
    return {
        "code": current[0],
        "points": round(points, 1),
        "next_code": nxt[0] if nxt else None,
        "points_to_next": round(to_next, 1),
        "progress": round(min(1.0, max(0.0, progress)), 3),
    }

File: backend-java/src/main/java/ru/savefood/gamification/Gamification.java
package ru.savefood.gamification;
import java.util.LinkedHashMap;
import java.util.Map;
public final class Gamification {
    private Gamification() {
    }
    private static final double POINTS_PER_DELIVERY = 10.0;
    private static final double POINTS_PER_KG = 1.0;
    /** (code, points threshold) ascending; i18n labels live on the frontend. */
    private static final String[] LEVEL_CODES = {"novice", "helper", "courier", "guardian", "city_hero"};
    private static final int[] LEVEL_THRESHOLDS = {0, 50, 200, 600, 1500};
    public static double computePoints(int deliveries, double kg) {
        return deliveries * POINTS_PER_DELIVERY + kg * POINTS_PER_KG;
    }
    /** Current level + progress towards the next one (for the progress bar). */
    public static Map<String, Object> computeLevel(int deliveries, double kg) {
        double points = computePoints(Math.max(0, deliveries), Math.max(0.0, kg));
        int currentIdx = 0;
        Integer nextIdx = null;
        for (int i = 0; i < LEVEL_CODES.length; i++) {
            if (points >= LEVEL_THRESHOLDS[i]) {
                currentIdx = i;
                nextIdx = i + 1 < LEVEL_CODES.length ? i + 1 : null;
            }
        }
        double progress;
        double toNext;
        if (nextIdx == null) {
            progress = 1.0;
            toNext = 0.0;
        } else {
            double span = LEVEL_THRESHOLDS[nextIdx] - LEVEL_THRESHOLDS[currentIdx];
            progress = span > 0 ? (points - LEVEL_THRESHOLDS[currentIdx]) / span : 1.0;
            toNext = Math.max(0.0, LEVEL_THRESHOLDS[nextIdx] - points);
        }
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("code", LEVEL_CODES[currentIdx]);
        out.put("points", round(points, 1));
        out.put("next_code", nextIdx == null ? null : LEVEL_CODES[nextIdx]);
        out.put("points_to_next", round(toNext, 1));
        out.put("progress", round(Math.min(1.0, Math.max(0.0, progress)), 3));
        return out;
    }
    private static double round(double v, int places) {
        double factor = Math.pow(10, places);
        return Math.round(v * factor) / factor;
    }
}
